trackintel/preprocessing/triplegs.py

```diff
@@ -218,6 +239,26 @@ def _seperate_ids(row):
         axis=1,
     )
 
+    # fill the missing points and convert to multipoint
+    if add_geometry:
+        # fill coordinates for origin staypoints that are NaN
+        origin_nan_rows = trips[pd.isna(trips["origin_staypoint_id"])].copy()
+        trips.loc[pd.isna(trips["origin_staypoint_id"]), "origin_geom"] = origin_nan_rows.tpls.map(
+            # from tpls table, get the first point of the first tripleg for the trip
+            lambda x: tpls.loc[x[0], tpls.geometry.name].boundary[0]
```

Comment:
I liked your solution better, and I think it was only unreadable because of the `trip_idx` variable etc that I had earlier. Now, I got rid of the nested function and simply repeat the same step for the distination.